### PR TITLE
[GCS] Optimize Redis query performance with `table_name:keys` dictionary

### PR DESCRIPTION
**Pull request by liuxsh9**
## Why are these changes needed?

**Problem:** When enabling Redis fault tolerance in GCS, operations involving queries, such as `get all job info`, experience poor performance, and the issue worsens as Redis records increase.

**Root Cause:** Redis contains 8 types of tables (e.g., `JOB`, `NODE`, `ACTOR`) distinguished only by the prefix of their keys. When querying a specific table, it requires traversing all Redis records to find those with matching prefixes, leading to inefficiencies.

**Solution:** Create a `table_name:keys` dictionary to accelerate queries. For example, when querying `JOB`, first use it to find all relevant keys, then perform the `AsyncMultiGet`. This dictionary is constructed by traversing Redis at cluster startup (for fault recovery) and updated with each Redis write and delete operation.

**Performance Improvement:**

image

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/store_client/redis_store_client.cc [hunk 1]
```diff
@@ -469,20 +488,15 @@ Status RedisStoreClient::AsyncGetKeys(
...
+  std::vector<std::string> result;
+  for (const auto &key : table_keys_cache_[table_name]) {
+    if (absl::StartsWith(key, prefix)) {
       result.push_back(key);
     }
-    callback(std::move(result));
-  };
-  return scanner->ScanKeysAndValues(match_pattern, on_done);
+  }
+  callback(std::move(result));
```
Comment (rynewang) [inline src/ray/gcs/store_client/redis_store_client.cc:498]: end the mutex lock before calling `callback`

File: src/ray/gcs/store_client/redis_store_client.cc [hunk 2]
```diff
@@ -214,7 +225,15 @@ Status RedisStoreClient::AsyncBatchDelete(const std::string &table_name,
   for (auto &key : keys) {
     redis_keys.push_back(GenRedisKey(external_storage_namespace_, table_name, key));
   }
-  return DeleteByKeys(redis_keys, callback);
+  return DeleteByKeys(redis_keys, [this, table_name, keys, callback](int64_t cnt) {
+    if (callback != nullptr) {
+      callback(cnt > 0);
+    }
+    absl::MutexLock lock(&cache_mu_);
```
Comment (rynewang) [inline src/ray/gcs/store_client/redis_store_client.cc:232]: We need to delete (ditto for add) the keys from the cache *before* calling DeleteByKeys, and don't hold locks for it

File: src/ray/gcs/store_client/redis_store_client.cc [hunk 3]
```diff
@@ -179,25 +190,25 @@ Status RedisStoreClient::AsyncGetAll(
...
-  std::string match_pattern =
-      GenKeyRedisMatchPattern(external_storage_namespace_, table_name);
-  auto scanner = std::make_shared<RedisScanner>(
-      redis_client_, external_storage_namespace_, table_name);
-  auto on_done = [callback,
-                  scanner](absl::flat_hash_map<std::string, std::string> &&result) {
-    callback(std::move(result));
-  };
-  return scanner->ScanKeysAndValues(match_pattern, on_done);
+  absl::MutexLock lock(&cache_mu_);
```
Comment (rynewang) [inline src/ray/gcs/store_client/redis_store_client.cc:193]: only hold lock for the cache lookup

File: src/ray/gcs/store_client/redis_store_client.cc [hunk 4]
```diff
@@ -141,17 +113,56 @@ RedisStoreClient::RedisStoreClient(std::shared_ptr<RedisClient> redis_client)
...
+  }
 }
 
 Status RedisStoreClient::AsyncPut(const std::string &table_name,
                                   const std::string &key,
                                   const std::string &data,
                                   bool overwrite,
                                   std::function<void(bool)> callback) {
+  auto put_callback = [this, table_name, key, callback](bool result) {
+    if (result) {
```
Comment (rynewang) [inline src/ray/gcs/store_client/redis_store_client.cc:156]: move the cache insertion to before calling DoPut

File: src/ray/gcs/store_client/redis_store_client.cc [hunk 5]
```diff
@@ -141,17 +113,56 @@ RedisStoreClient::RedisStoreClient(std::shared_ptr<RedisClient> redis_client)
   RAY_CHECK(!absl::StrContains(external_storage_namespace_, kClusterSeparator))
       << "Storage namespace (" << external_storage_namespace_ << ") shouldn't contain "
       << kClusterSeparator << ".";
+  const int TABLE_COUNT = 8;
+  std::string tables[TABLE_COUNT] = {
```
Comment (rynewang) [inline src/ray/gcs/store_client/redis_store_client.cc:117]: make this a constexpr static variable. Also make a function to do the initial cache population

File: src/ray/gcs/store_client/redis_store_client.cc [hunk 6]
```diff
@@ -179,25 +190,25 @@ Status RedisStoreClient::AsyncGetAll(
...
-      callback(cnt > 0);
-    }
-  });
+  return AsyncBatchDelete(
+      table_name, {key}, [this, table_name, key, callback](int64_t cnt) {
+        if (callback != nullptr) {
+          callback(cnt > 0);
+        }
+        absl::MutexLock lock(&cache_mu_);
+        this->table_keys_cache_[table_name].erase(key);
```
Comment (rynewang) [inline src/ray/gcs/store_client/redis_store_client.cc:210]: since we already did that in AsyncBatchDelete, this is not needed?

**Comment by rynewang** [comment]
Some comments.

0. The first issue is that it assumes all modifications (add, remove) goes through this redis client, so it can capture all deltas in the cache. However in gcs_server we have [GcsTableStorage](https://github.com/ray-project/ray/blob/aaee46166ad65f04a3d34919720b219282c17c8b/src/ray/gcs/gcs_server/gcs_table_storage.h#L297) which holds a client; and [GcsInternalKVManager](https://github.com/ray-project/ray/blob/aaee46166ad65f04a3d34919720b219282c17c8b/src/ray/gcs/gcs_server/gcs_server.cc#L571) which holds another client. This *should be* fine since KV manager only ever reads/writes KV table and GcsTableStorage does not read/write KV at all. But maybe we need to audit a bit more to make sure things are safe. A quick search of `RedisStoreClient` in the whole repo suggests we only have these 2 meaningful places to init, so ok.
1. This effectively caches *all* keys in Redis in GCS. Is it gonna cost big memory usage in `gcs_server`? I think it should be fine, since most tables are keyed by [Ray IDs](https://github.com/ray-project/ray/blob/master/src/ray/design_docs/id_specification.md) (TaskID is 24 bytes and we only have actor creation tasks); but Internal KV keys may be unbounded in count and in length. @jjyao to confirm on possible mem usage
2. Looks like the code of Redis API callings and the cache code are entangled. Can you make another subclass of `class CachingKeyStoreClient : public StoreClient` that internally inits a `std::unique_ptr<StoreClient>` and keeping the cache?
3. For the order of cache vs redis: Cache updating must happen before calling user callback, but it's ok to put it either at beginning of the Async func call, or at the beginning in the async callback. I prefer to put it right at the beginning of each call.
4. Locks should only lock the cache read/write, not the user callbacks.

**Comment by rynewang** [comment]
hmm I also found we already have this optimization named `GcsTableWithJobId`. Just that it's not used for `GcsJobTable` (get_all_job_info). Can you try to see if it can be used for your desired APIs?

**Comment by liuxsh9** [comment]
Thanks for the review and excellent suggestions! @rynewang I will check it soon.

**Comment by liuxsh9** [comment]
Thanks for the review. 
SomeUpdates and feedbacks:

0. Holding two clients may be not a big issue, but cache might waste some memory, especially initializated from large Redis records. We're trying to merge them - what do you think?
1. Have tested with 1.5 million Redis records, the new cache used around 100MB memory. However, we didn't consider unbounded internal KV keys. Any suggestions, @jjyao ? Thanks!
2. Have created a new `CacheKeyStoreClient` subclass, which helps decouple cache logic. Also tried to add a cache enable switch for more flexibility.
3. We concerned about cache coherency on Redis op (ex. `PUT`) failures. So we keep to execute the cache op after the Redis op is successful.
4. Fixed the locking issue, thanks for pointing it out!
5. `GcsTableWithJobId` can only be used for `ACTOR`, `ACTOR_TASK_SPEC` and `JOB` due to they are the only ones with the job ID field. Querying other tables remains slow, does not seem to work directly.

Please review these updates. Thanks!

**Comment by jjyao** [comment]
@liuxsh9 @rynewang let's explore to see if we can have a redis Hash per gcs table. Currently we have a single Hash storing all the data.

**Comment by rynewang** [comment]
@liuxsh9 I assembled this PR that changes schema of how we use redis - now we should have 1 hash per table, so get all keys becomes a HSCAN that only scans that hash. Would you mind do some benchmarking on that? Thanks. https://github.com/ray-project/ray/pull/46861

**Comment by jjyao** [comment]
@liuxsh9 In practice, do you see this as a performance issue? Your perf shows 1.5 million records in redis, do you actually have a Ray cluster with that many of records?

**Comment by liuxsh9** [comment]
> @liuxsh9 In practice, do you see this as a performance issue? Your perf shows 1.5 million records in redis, do you actually have a Ray cluster with that many of records?

@jjyao Yes, when we using a persistent Ray cluster with over a dozen nodes, and after a week, the number of records in Redis exceeds 1.5 million. At this point, all operations that use the `GetAllJobInfo` interface will become slow. These records are mainly `worker` records, because `worker` currently don't have a maximum number limit and a cleanup mechanism like `actor` do. Perhaps we can contribute to this part?

**Comment by liuxsh9** [comment]
@rynewang https://github.com/ray-project/ray/pull/46861 This is great! I think it solves the problem from a more fundamental level, so we no longer need to use caching.

image

**Comment by jjyao** [comment]
@liuxsh9 #46861 is merged. Please try it out.

**Comment by liuxsh9** [comment]
> @liuxsh9 #46861 is merged. Please try it out.

We've tested the performance and seen a clear improvement, meeting our expectations. Thanks!

**Pull request closed without merging**